HW3/P5/label_regions.cl

```diff
@@ -80,6 +82,28 @@ propagate_labels(__global __read_write int *labels,
     old_label = buffer[buf_y * buf_w + buf_x];
 
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
+    int buf_x_i, buf_y_i;
+    int label = -1;
+    int idx = -1;
+    
+    if (idx_1D == 0) {
+        
+        for (buf_x_i = halo; buf_x_i < buf_x - halo; buf_x_i++) {
+            
+            for (buf_y_i = halo; buf_y_i < buf_h - halo; buf_y_i++) {
+                
+                if (buffer[buf_y_i * buf_w + buf_x_i] == idx) {
+                    buffer[buf_y_i * buf_w + buf_x_i] = label;
+                }
+                else if (buffer[buf_y_i * buf_w + buf_x_i] < w * h) {
+                    idx = buffer[buf_y_i * buf_w + buf_x_i];
+                    label = labels[idx];
```

Comment:
It seems like you are missing the update of buffer[buf_y_i * buf_w + buf_x_i] to the value of the grandparent.